File: kernel/kernel-api/src/main/java/io/delta/kernel/commit/actions/AbstractMetadata.java
/*
 * Copyright (2021) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.delta.kernel.commit.actions;

import java.util.List;
import java.util.Map;

/**
 * Interface for metadata actions in Delta. The metadata defines the metadata
 * of the table.
 */
public interface AbstractMetadata {

    /**
     * A unique table identifier.
     */
    String getId();

    /**
     * User-specified table identifier.
     */
    String getName();

    /**
     * User-specified table description.
     */
    String getDescription();

    /** The table provider format. */
    String getProvider();

    /** The format options */
    Map<String, String> getFormatOptions();

    /**
     * The table schema in string representation.
     */
    String getSchemaString();

    /**
     * List of partition columns.
     */
    List<String> getPartitionColumns();

    /**
     * The table properties defined on the table.
     */
    Map<String, String> getConfiguration();

    /**
     * Timestamp for the creation of this metadata.
     */
    Long getCreatedTime();
}

File: spark/src/main/scala/org/apache/spark/sql/delta/v2/interop/AbstractMetadata.scala
/*
 * Copyright (2025) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.delta.v2.interop

import org.apache.spark.sql.types.StructType

/**
 * Abstract trait for metadata actions in Delta. This trait provides a common
 * abstraction that can be implemented by both Spark's V1 Metadata and Kernel's MetadataV2
 * in V2 connector. The V2 connector will implement adapters for reusing V1 utilities.
 */
trait AbstractMetadata {

  /** A unique table identifier. */
  def id: String

  /** User-specified table identifier. */
  def name: String

  /** User-specified table description. */
  def description: String

  /** Returns the schema as a [[StructType]]. */
  def schema: StructType

  /** List of partition column names. */
  def partitionColumns: Seq[String]

  /** The table properties/configuration defined on the table. */
  def configuration: Map[String, String]
}

